### ISSUE TITLE
New SDK not working on 2K display phones !!!

### ISSUE THREAD

**Issue by WTF-Grinch** [googlesamples/cardboard-unity#88]
when building with using new SDK on LG G4 screen size is not changed and it is fitted so when you put it in Cardboard image is blurred and not focused.

**Comment by smdol** [comment]
Can you post a screenshot?

**Comment by WTF-Grinch** [comment]
![11837092_10156043864895316_462726821_o](https://cloud.githubusercontent.com/assets/2622331/9977314/225e8fe6-5f13-11e5-8f68-2927d6afabdd.jpg)

**Comment by WTF-Grinch** [comment]
it is on LG G4

**Comment by WTF-Grinch** [comment]
i have tested SDK cardboard-unity-0.4.10 and no problem

**Comment by smdol** [comment]
Is it cut off on the right side like that on your screen, or is that just the screenshot?

Also, the function that determines the screen size is in BaseCardboardDevice.cs, called Init().  It passes several numbers like Screen.width, Screen.dpi, etc, into the native Start() function.  Could you log those numbers and post them here?

(In the 0.4.10 SDK, screen resolution and size were determined by Java code, which used a different method.)

**Comment by WTF-Grinch** [comment]
width is 2560 
height is 1440 
DPI 640

**Comment by WTF-Grinch** [comment]
Has Anyone some solution or maybe fix ?

**Comment by WTF-Grinch** [comment]
Also old one has some problem when loading multi levels screen goes black :( i am using new one for my project but on 2K screens the problem is same so it will be good if developers will make some sort of fix. :)

**Comment by smdol** [comment]
I'm looking for an LG G4 to try this out on.  If I find one, I'll see what is going on.

About the scene going black on level load, there were some bugs in the old SDK depending on how you used the Cardboard scripts in the various levels.  So yes, this definitely needs to get fixed :)

**Comment by WTF-Grinch** [comment]
it will be great :) i will wait our feedback maybe i am doing something wrong

**Comment by WTF-Grinch** [comment]
any updates ?

**Comment by WTF-Grinch** [comment]
? ?

**Comment by smdol** [comment]
Unable to reproduce the problem so far on an LG G4 in house.  Seems to work fine with the latest release, and either Unity 4 or Unity 5.

![screenshot_thu](https://cloud.githubusercontent.com/assets/10277940/10119623/e61104c8-644f-11e5-97d1-7104180acda7.png)

**Comment by WTF-Grinch** [comment]
can you share your settings that you have on cardboard prefab and player pref with quality ?

**Comment by smdol** [comment]
For quality, I just kept the default settings for Android that you get when you start a new project.  Similarly, the default settings on the prefab that you get when you add it to the scene.  I built the apk and gave it to a coworker who has an LG G4 to test, and the above screenshot is the result.

This is a strange error, if your screenshot is correct:  I usually see issues where the eyes are too small but centered horizontally, or the eyes are separated too far (but still centered).  Both happen when the phone returns an inaccurate DPI.  Being cut off on the right like your photo shows is a new one.

**Comment by WTF-Grinch** [comment]
hmmmi have the same i did not changed anything everything is ok but when i put it in the cardboard the image is not in focus it is not centered and in your case view is smaller then mine. this black rectangles are bigger in my case i will upload new image maybe old one is bad :(

**Comment by smdol** [comment]
Which Cardboard viewer do you have?  You didn't have any trouble scanning the QR code, right?

**Comment by WTF-Grinch** [comment]
2015 Google Cardboard V2.0 and no i have no problems with QR code

**Comment by WTF-Grinch** [comment]
![12062363_10156074485030316_416204293_o](https://cloud.githubusercontent.com/assets/2622331/10168342/679befbe-66dd-11e5-9aed-e50123256e11.jpg)

**Comment by smdol** [comment]
What version of Unity are you using?  

I just saw a comment on another thread that different versions of Unity (5.2 vs 5.1.3) are giving different DPI values on the same phone.  From your photo, I'm guessing that the DPI value Unity is returning is too high.

(BTW, in SDK 0.4.10 and earlier, DPI was retrieved in the Java layer, not through Unity, so this problem would not occur.)

**Comment by smdol** [comment]
Found some code on the Unity forums that should give the same results as the 0.4.10 and earlier SDK would give:

http://forum.unity3d.com/threads/finding-physical-screen-size.203017/#post-1373867

Give this a try and see if it fixes the problem.

**Comment by WTF-Grinch** [comment]
I am using 5.1.3p2 and DPI is 640 i will check with other version of unity. i have searched same thread i will try that too :)

**Comment by momo-the-monster** [comment]
I've got a similar issue. I'm deploying to a Galaxy Note 4. All the existing demos are fine, but after updating to Unity 5.2, builds from Unity just don't look right.

Screen.dpi reports 640, which is incorrect. I calculated my DPI to be 515 (2560x1440 at 5.7") and hardcoded that into BaseCardboardDevice::Init and things look right again!

Screenshot below - Example scene from Cardboard Demos up top, which looks correct in the Cardboard Viewer, Example Unity Scene below, which looks real bad in the viewer.

![cardboard-dpi-issue](https://cloud.githubusercontent.com/assets/737888/10236279/9a795fbc-6858-11e5-9bcf-a08816f7069c.jpg)

**Comment by smdol** [comment]
Yep, I'm hearing this a lot, and we even reproduced it in house.  I don't know if this is because of a bug in Unity or a policy change.  It's on the bug list to be fixed now for sure.

**Comment by momo-the-monster** [comment]
Glad to hear it, thanks for the update!

**Comment by WTF-Grinch** [comment]
Before new fix will arrive as @smdol sujested you can create new script in unity add this to class 

public class DisplayMetricsAndroid {
 
    // The logical density of the display
    public static float Density { get; protected set; }
 
    // The screen density expressed as dots-per-inch
    public static int DensityDPI { get; protected set; }
 
    // The absolute height of the display in pixels
    public static int HeightPixels { get; protected set; }
 
    // The absolute width of the display in pixels
    public static int WidthPixels { get; protected set; }
 
    // A scaling factor for fonts displayed on the display
    public static float ScaledDensity { get; protected set; }
 
    // The exact physical pixels per inch of the screen in the X dimension
    public static float XDPI { get; protected set; }
 
    // The exact physical pixels per inch of the screen in the Y dimension
    public static float YDPI { get; protected set; }
 
    static DisplayMetricsAndroid() {
       // Early out if we're not on an Android device
       if (Application.platform != RuntimePlatform.Android) {
         return;
       }
 
       // The following is equivalent to this Java code:
       //
       // metricsInstance = new DisplayMetrics();
       // UnityPlayer.currentActivity.getWindowManager().getDefaultDisplay().getMetrics(metricsInstance);
       //
       // ... which is pretty much equivalent to the code on this page:
       // http://developer.android.com/reference/android/util/DisplayMetrics.html
 
       using (
         AndroidJavaClass unityPlayerClass = new AndroidJavaClass("com.unity3d.player.UnityPlayer"),
         metricsClass = new AndroidJavaClass("android.util.DisplayMetrics")
       ) {
         using (
          AndroidJavaObject metricsInstance = new AndroidJavaObject("android.util.DisplayMetrics"),
          activityInstance = unityPlayerClass.GetStatic<AndroidJavaObject>("currentActivity"),
          windowManagerInstance = activityInstance.Call<AndroidJavaObject>("getWindowManager"),
          displayInstance = windowManagerInstance.Call<AndroidJavaObject>("getDefaultDisplay")
         ) {
          displayInstance.Call("getMetrics", metricsInstance);
          Density = metricsInstance.Get<float>("density");
          DensityDPI = metricsInstance.Get<int>("densityDpi");
          HeightPixels = metricsInstance.Get<int>("heightPixels");
          WidthPixels = metricsInstance.Get<int>("widthPixels");
          ScaledDensity = metricsInstance.Get<float>("scaledDensity");
          XDPI = metricsInstance.Get<float>("xdpi");
          YDPI = metricsInstance.Get<float>("ydpi");
         }
       }
    }
}

and in BaseCardboardDevice.cs search for init function and replace with this one :

	public override void Init() {
		// Get landscape dimensions, even if app starts in portrait.
		int width = Mathf.Max(Screen.width, Screen.height);
		int height = Mathf.Min(Screen.height, Screen.width);

		float xdpi = Screen.dpi;
		float ydpi = Screen.dpi;

#if UNITY_ANDROID && !UNITY_EDITOR
		xdpi = DisplayMetricsAndroid.XDPI;
		ydpi = DisplayMetricsAndroid.YDPI;
#endif

		Start(width, height, xdpi, ydpi);
		SetEventCallback(OnVREvent);
	}

for me it is working. i do not have iphone 6 to test but if there will be same problem you can do same for iphone to call native function that will return dpi from Objective c. 

@smdol  thank you one more time dude :)

**Comment by noemis84** [comment]
I had same problem on samsung s6 after upgrade to unity 5.2 - this fix solves it.

Thank you @WTF-Grinch @smdol  !

**Comment by smdol** [comment]
New v0.5.2 release should fix these weird DPI issues which depend on the version of Unity.

**Comment by billyjoecain** [comment]
Thank you for posting this. We created 7 apps that seem to work perfectly and on our 8th, we started seeing bizarre results. Our developer found this post after we encountered our issues. They seem to be exhibiting the same issue. In our most recent app, we have two scenes, one with a logo and preparatory sequence and a second one with the main experience. 

Scene one is almost always 'off' and Scene two is randomly 'right' or 'wrong.' We could not find any reason for this.

It is happening on our current version of Unity, 5.2.1f1.
